compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -147,6 +147,12 @@ pub(crate) fn try_destructure_const<'tcx>(
     // We go to `usize` as we cannot allocate anything bigger anyway.
     let (field_count, variant, down) = match val.ty().kind() {
         ty::Array(_, len) => (usize::try_from(len.eval_usize(tcx, param_env)).unwrap(), None, op),
+        // Checks if we have any variants, to avoid downcasting to a non-existing variant (when
+        // there are no variants `read_discriminant` successfully returns a non-existing variant
+        // index).
+        ty::Adt(def, _) if def.variants.is_empty() => {
+            return Ok(mir::DestructuredConst { variant: None, fields: &[] });
```

Comment:
If we can't avoid this special case, then it would be much better to return an `Err`. Making up a fake value seems bad.

Also, I don't know what `def.variants` will be like for a union, so it might be better to ensure this is actually an enum?

Reply:
Now that we have a `Result` return type, `Err` does indeed would be a much better choice.

Unions are single variant with overlapping fields. An enum is the only `Adt` that might not have any variants.